### PR TITLE
Copter: motor spool changes part 2

### PR DESCRIPTION
**Pull request by rmackay9**
This PR builds on [this earlier PR from bnsgeyer that added spool logic to the AP_Motors library for TradHelis](https://github.com/ArduPilot/ardupilot/pull/10212) and [replaces this earlier PR](https://github.com/ArduPilot/ardupilot/pull/7894).

The changes in this PR are:

- TradHeli replaces some calls to **rotor_runup_complete** with checks of the motor spool state
- Flight mode state logic is consolidated across many modes
- make_safe_shut_down function is added and used instead of immediately calling init_disarm().  The purpose is to maintain attitude control while the propellers are slowing down which is especially important on tradhelis
- calls to attitude_control::set_throttle_out_unstabilized (which simply turned off attitude control) are replaced with calls that attempt to maintain attitude control as long as possible.
- knock-on impacts to Plane and Sub are also included although these are mostly non-functional changes
- minor formatting fixes

As a side note, the commits are probably over-squashed.  I could probably split out the "Copter: consolidate mode state decisions" into multiple commits if that would help with the review.

There are a lot of changes in this PR so although it passes Travis it will need careful review before merging.

Contributors to this PR are @bnsgeyer, @lthall and some help from @rmackay9.

### CODE REVIEW

File: ArduCopter/mode.cpp [hunk 1]
```diff
@@ -380,14 +376,39 @@ void Copter::Mode::zero_throttle_and_relax_ac(bool spool_up)
...
+}
+
+void Copter::Mode::zero_throttle_and_hold_attitude()
+{
+    // run attitude controller
+    attitude_control->input_rate_bf_roll_pitch_yaw(0.0f, 0.0f, 0.0f);
+    attitude_control->set_throttle_out(0.0f, false, copter.g.throttle_filt);
+}
+
+void Copter::Mode::make_safe_shut_down()
```
Comment (bnsgeyer) [inline ArduCopter/mode.cpp]: @rmackay9  I noticed in make_safe_shutdown that the init_disarm_motors was no longer called from here.  the last version that we worked on had the disarm called from this method.  Why was this changed?

File: ArduCopter/mode_rtl.cpp [hunk 2]
```diff
@@ -358,20 +353,18 @@ bool Copter::ModeRTL::landing_gear_should_be_deployed() const
...
-        loiter_nav->init_target();
+    // check if we've completed this stage of RTL
+    _state_complete = ap.land_complete;
 
-        // disarm when the landing detector says we've landed
-        if (ap.land_complete && disarm_on_land) {
-            copter.init_disarm_motors();
-        }
+    // disarm when the landing detector says we've landed
+    if (disarm_on_land && ap.land_complete) {
```
Comment (bnsgeyer) [inline ArduCopter/mode_rtl.cpp]: So just taking mode_rtl as an example.  Since init_disarm_motors is called from here, then in order to let it run make_safe_shut_down then we have to add a condition to this statement that requires the spool mode to be Ground Idle before allowing disarm.  Is this the way forward? if so, I can go through and make the changes to have it disarm properly in these cases.

**Comment by bnsgeyer** [comment]
@rmackay9 looking at the auto mode again, my change didn’t completely fix the issue. It seems that the system exits the RTL and Land routines prior to allowing the spool state to get to ground idle. 
I guess a more universal question for auto mode is what is the expectation for the spool state after a landing?  I am assuming we just want it to go to ground idle and not disarm. It appears to me when I look at the code that a disarm is only completed when it exits the mission. I want to be sure I have this interpreted correctly before marking changes to this end.

**Comment by rmackay9** [comment]
@bnsgeyer, yes, I think that the vehicle should only disarm once the mission ends.  I think after landing/rtl-ing it should spool-down to ground-idle.  Perhaps we don't consider the RTL or Land complete until it has spooled down.

**Comment by bnsgeyer** [comment]
@rmackay9 I wanted to post the work that I've been doing to test the spool logic.  Here is a link to a powerpoint file that contains a summary of the testing and how it was tested.  I still have a few more tests but I wanted to let  you see this.  Let me know if there are other tests you would like to see.
[spool logic testing presentation](https://drive.google.com/open?id=19wIGuXLBRPDsqpsncXgSP4TH3aTH4ZS-)

**Comment by rmackay9** [comment]
@bnsgeyer, thanks for the document and it's great to see this thorough testing.  I've looked through it and some questions/answers:

- "Once landed aircraft spools down without issue but does not shutdown-IS THIS DESIRED?".  Yes, I think it is good that a companion computer is able to land the vehicle and then takeoff again.  If the companion computer really wants to shutdown the vehicle they could send a disarm command I think.
- "Poshold Roll Axis: Climb rate still controllable after motor interlock disabled".  Is this a good thing or bad?  sorry, I'm not sure but I think maybe it's a good thing?

So I guess you're generally happy with the spool logic now?

**Comment by rmackay9** [comment]
I've created a rebased version of this PR here in [bnsgeyer-spool-logic10](https://github.com/rmackay9/rmackay9-ardupilot/commits/bnsgeyer-spool-logic10).  I will force-push it over this branch after we're sure the rebase is OK.  In particular there is a question about a check in set_mode() that I've ping'd over to @bnsgeyer.

**Comment by bnsgeyer** [comment]
@rmackay9  I reviewed the conflicts and how you handled them and I had no issues with it.  As for the issue you raised with me on skype, that is the desired behavior.  Looking back at the history, it appears that Rob had made this change back in 3.3 and Tridge pulled it into the mode function for  heli's.  How it got combined to include multi's I'm not sure.  The one thing I noticed is that it doesn't allow the user to switch to a non-throttle mode on spool down either.  I don't think this is a big deal, if the rotor is spooling down and the heli is still flying, the user should not be switching into an non-manual throttle mode. That would surely end up in disaster.  So I think the logic here is sound for heli's.  The only thing I would bring up is whether you want to get rid of the #if frame statement and replace with if (is_tradheli()) statement.

So I approve of bnsgeyer-spool-logic10 branch.  thanks for all of your hard work!

**Comment by rmackay9** [comment]
I've test flown this on my quad and it seems fine.  I only tested Stabilize, Loiter and Auto but didn't notice any real differences vs master.

**Comment by kd0aij** [comment]
@rmackay9 The set_throttle_out changes look OK to me... Flies OK with the Stryker_quad in RF8 SITL for QSTABILIZE, QHOVER and QLOITER modes. 
@IamPete1 could you test a regular qudplane?

**Comment by IamPete1** [comment]
will do

**Comment by IamPete1** [comment]
works fine for standard quadplane in real flight

**Comment by rmackay9** [comment]
rebased on master..

**Comment by rmackay9** [comment]
One change in behaviour we have is that if there is a LAND waypoint in the middle of the mission, it must be followed by a TAKEOFF command or the mission will become stuck.  We can probably add this to the wiki and perhaps someday we should add a mission-check feature that ensures that users don't hit problems like this.

**Comment by rmackay9** [comment]
I've tested the spool-up/spool-down logic in SITL for acro, althold, auto, brake, circle, drift, flip, flowhold, guided (position, velocity and attitude control but not posvel), land, loiter, poshold, rtl, sport and stabilize and I'm pretty sure that these modes are working correctly.

From a visual inspection follow also looks fine although I was unable to test follow and throw because those are a bit tricky in the simulator.

Some remaining question/comments:

- Flip has an out-of-date comment, "// Don't enter THROW mode if interlock will prevent motors running".  it looks like we've fixed an existing bug by removing the "interlock" check at the top of run()?
- Flip replaces "set_throttle_out_unstabilized()" with a call to "set_throttle_out()".  Don't we also need to add, "attitude_control->relax_attitude_controllers();"?  If not we should probably fix this comment, "// Hold throttle at zero during the throw and continually reset the attitude controller"
- Follow: the new set-desired-spool-state(UNLIMITED) has been added in this PR but is probably not necessary because there is a duplicate call from within guided mode's velocity controller.
- One time when landing with Guided's velocity controller the vehicle flipped over.  I could not reproduce the issue though so I think it was a one-off

**Comment by WickedShell** [comment]
> One change in behaviour we have is that if there is a LAND waypoint in the middle of the mission, it must be followed by a TAKEOFF command or the mission will become stuck.

This is presumably only a problem if you complete the landing waypoint right? I frequently witness people start waypoints (particularly landing ones), then change their mind and redirect the aircraft to a different waypoint.

File: ArduCopter/mode_auto.cpp [hunk 3]
```diff
@@ -872,9 +865,10 @@ void Copter::ModeAuto::nav_guided_run()
 //      called by auto_run at 100hz or more
 void Copter::ModeAuto::loiter_run()
 {
-    // if not auto armed or motor interlock not enabled set throttle to zero and exit immediately
-    if (!motors->armed() || !ap.auto_armed || ap.land_complete || !motors->get_interlock()) {
-        zero_throttle_and_relax_ac();
+    // if not armed set throttle to zero and exit immediately
+    if (!motors->armed() || !ap.auto_armed || ap.land_complete) {
```
Comment (khancyr) [inline ArduCopter/mode_auto.cpp]: maybe we could factorize this block, it is use 3 time

File: ArduCopter/mode_circle.cpp [hunk 4]
```diff
@@ -27,41 +27,30 @@ bool Copter::ModeCircle::init(bool ignore_checks)
...
-        // check for pilot requested take-off
-        if (ap.land_complete && target_climb_rate > 0) {
-            // indicate we are taking off
-            set_land_complete(false);
-            // clear i term when we're taking off
-            set_throttle_takeoff();
-        }
+    // if not armed set throttle to zero and exit immediately
+    if (!motors->armed() || !ap.auto_armed || ap.land_complete) {
+        make_safe_spool_down();
```
Comment (khancyr) [inline ArduCopter/mode_circle.cpp]: why not copter.circle_nav->init(); ?

File: ArduCopter/mode_drift.cpp [hunk 5]
```diff
@@ -64,34 +52,56 @@ void Copter::ModeDrift::run()
...
+        // Motors should be Stopped
+        motors->set_desired_spool_state(AP_Motors::DESIRED_SHUT_DOWN);
+    } else if (ap.throttle_zero) {
+        // Attempting to Land
+        motors->set_desired_spool_state(AP_Motors::DESIRED_GROUND_IDLE);
+    } else {
+        motors->set_desired_spool_state(AP_Motors::DESIRED_THROTTLE_UNLIMITED);
+    }
+
+    if (motors->get_spool_mode() == AP_Motors::SHUT_DOWN) {
```
Comment (khancyr) [inline ArduCopter/mode_drift.cpp]: can't this be a switch case ?

File: ArduCopter/mode_acro_heli.cpp [hunk 6]
```diff
@@ -29,24 +29,36 @@ void Copter::ModeAcro_Heli::run()
...
+        motors->set_desired_spool_state(AP_Motors::DESIRED_SHUT_DOWN);
+    } else {
+        // heli will not let the spool state progress to THROTTLE_UNLIMITED until motor interlock is enabled
+        motors->set_desired_spool_state(AP_Motors::DESIRED_THROTTLE_UNLIMITED);
     }
 
-    // clear landing flag above zero throttle
-    if (motors->armed() && motors->get_interlock() && motors->rotor_runup_complete() && !ap.throttle_zero) {
-        set_land_complete(false);
+    if (motors->get_spool_mode() == AP_Motors::SHUT_DOWN) {
```
Comment (khancyr) [inline ArduCopter/mode_acro_heli.cpp]: can't this be a switch case ?

File: ArduCopter/mode_acro.cpp [hunk 7]
```diff
@@ -10,21 +10,34 @@
...
 {
+    // convert the input to the desired body frame rate
     float target_roll, target_pitch, target_yaw;
+    get_pilot_desired_angle_rates(channel_roll->get_control_in(), channel_pitch->get_control_in(), channel_yaw->get_control_in(), target_roll, target_pitch, target_yaw);
 
-    // if not armed set throttle to zero and exit immediately
-    if (!motors->armed() || ap.throttle_zero || !motors->get_interlock()) {
-        zero_throttle_and_relax_ac();
-        return;
+    if (!motors->armed()) {
```
Comment (khancyr) [inline ArduCopter/mode_acro.cpp]: this block is use numerous time, we could fatorize it

**Comment by rmackay9** [comment]
@WickedShell, thanks for the comments.  Yes, the requirement for a takeoff (after a Land command) is only if the vehicle successfully lands and spools down the motors.  If the user aborts the land somehow (for example advances the mission before the land command completes) then there's no need for the takeoff.

File: ArduCopter/mode_follow.cpp [hunk 8]
```diff
@@ -26,12 +26,15 @@ bool Copter::ModeFollow::init(const bool ignore_checks)
 
 void Copter::ModeFollow::run()
 {
-    // if not auto armed or motor interlock not enabled set throttle to zero and exit immediately
-    if (!motors->armed() || !ap.auto_armed || !motors->get_interlock()) {
-        zero_throttle_and_relax_ac();
+    // if not armed set throttle to zero and exit immediately
+    if (!motors->armed() || !ap.auto_armed || ap.land_complete) {
+        make_safe_spool_down();
```
Comment (khancyr) [inline ArduCopter/mode_follow.cpp]: wp_nav->wp_and_spline_init(); ?

File: ArduCopter/mode_auto.cpp (hunk 3 above)
Comment (rmackay9) [reply ArduCopter/mode_auto.cpp]: Yes we probably could although it's an existing issue ([see master here](https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/mode_auto.cpp#L876)) so I don't think I will increase the scope of this PR any more.

File: ArduCopter/mode_acro_heli.cpp (hunk 6 above)
Comment (khancyr) [reply ArduCopter/mode_acro_heli.cpp]: in fact the whole block can be factorize as use multiple time accross modes

File: ArduCopter/mode_drift.cpp (hunk 5 above)
Comment (rmackay9) [reply ArduCopter/mode_drift.cpp]: Yes, perhaps it could be although there are two enum values that are missing (perhaps they should be added too but I'm not sure).. I'll check with @lthall.  BTW,  the same applies for Acro and Stabilize mode.

Txs for the review and comments!

File: ArduCopter/mode_acro_heli.cpp (hunk 6 above)
Comment (rmackay9) [reply ArduCopter/mode_acro_heli.cpp]: possibly, basically it's the same pattern as drfit...

File: ArduCopter/mode_circle.cpp (hunk 4 above)
Comment (rmackay9) [reply ArduCopter/mode_circle.cpp]: Thanks again for the review - it's very much appreciated because this is quite a bit RP.

So we are discussing what should happen when the vehicle is landed in Circle mode.  This is a real edge case and could only happen if:

- the user lands the vehicle by pulling down on the throttle stick
- the vehicle was landed in some other mode (like Loiter) and the user then switches to Circle

Calling circle_nav->init() would cause the center of the circle to be reset depending upon the vehicle's speed (it's not moving so it would be directly ahead of the vehicle).  It's not totally obvious if this is what we should do - if the user has not taken off yet then I agree it's probably the correct thing to do.  If the user has already been flying in Circle and has landed and is then taking off again then they might not expect the circle center to be moved.

In either case though, we don't support taking off in Circle mode.  Even if the users pulls the throttle stick up, the vehicle will not take off so in the end, they will be forced to switch to another mode, takeoff and then switch back into circle mode which will cause circle_nav->init() to be called.

File: ArduCopter/mode_acro.cpp (hunk 7 above)
Comment (rmackay9) [reply ArduCopter/mode_acro.cpp]: yes, i agree but I'll leave that out-of-scope for the moment if that's OK because it's an existing issue in master and this PR is already scary enough :-)

**Comment by khancyr** [comment]
@rmackay9 no worries, the changes seems good. We could factorize after this PR is merged.
the only point that bother me is on the block
````
if (!motors->armed() || !ap.auto_armed || ap.land_complete) {
````
sometime you reinit the controler (loiter/wp etc) sometime you don't. I don't understand the inconsistancy

File: ArduCopter/mode_follow.cpp (hunk 8 above)
Comment (rmackay9) [reply ArduCopter/mode_follow.cpp]: It wouldn't be the wp_and_spline_init() but it could be that we should be calling pos_control->init_vel_controller_xyz();


I think you're probably correct that it is a problem but I'm probably not going to address it as part of this PR because in eithe case it would be an existing problem because we don't call the init in master either.  The way this problem would show up is as a very sudden start of the vehicle.  I.e. it wouldn't be acceleration limited I think.

**Comment by rmackay9** [comment]
I've added a new commit after discussing with Leonard which addresses the comments I made above regarding throw mode.  If there are no further comments I'll probably merge this to master no later than Monday.
